src/json_rpc.rs

```diff
@@ -0,0 +1,116 @@
+use futures::TryFutureExt;
+use reqwest::Url;
+use serde::{de::DeserializeOwned, Serialize};
+use std::fmt::Debug;
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Clone, Debug)]
+pub struct Client {
+    inner: reqwest::Client,
+    url: Url,
+}
+
+impl Client {
+    pub fn new(base_url: Url) -> Self {
+        Self {
+            inner: reqwest::Client::new(),
+            url: base_url,
+        }
+    }
+
+    pub async fn send<Req, Res>(&self, request: Request<Req>) -> Result<Res>
+    where
+        Req: Debug + Serialize,
+        Res: Debug + DeserializeOwned,
+    {
+        self.send_with_path("".into(), request).await
+    }
+
+    pub async fn send_with_path<Req, Res>(&self, path: String, request: Request<Req>) -> Result<Res>
+    where
+        Req: Debug + Serialize,
+        Res: Debug + DeserializeOwned,
+    {
+        let url = self.url.clone().join(&path)?;
+
+        let response = self
+            .inner
+            .post(url.clone())
+            .json(&request)
+            .send()
+            .map_err(Error::ConnectionFailed)
+            .await?;
+
+        let response = response.bytes().await?;
+
+        let response: Response<Res> = match serde_json::from_slice(&response) {
+            Ok(response) => response,
+            Err(error) => {
+                let response = String::from_utf8_lossy(&response[..]);
+                tracing::debug!("Response received: {}", response);
+                return Err(Error::Serde(error));
+            }
+        };
+
+        match response {
+            Response::Success { result } => Ok(result),
+            Response::Error { error } | Response::RpcError(error) => Err(Error::JsonRpc(error)),
+        }
+    }
+}
+
+#[derive(serde::Serialize, Debug, Clone)]
+pub struct Request<T> {
+    id: String,
+    jsonrpc: String,
+    method: String,
+    params: T,
+}
+
+impl<T> Request<T> {
+    pub fn new(method: &str, params: T, jsonrpc: String) -> Self {
+        Self {
+            id: "1".to_owned(),
+            jsonrpc,
+            method: method.to_owned(),
+            params,
+        }
+    }
+}
+
+#[derive(serde::Deserialize, Debug)]
+#[serde(untagged)]
+pub enum Response<T> {
+    Success { result: T },
+    Error { error: JsonRpcError },
+    RpcError(JsonRpcError),
+}
```

Comment:
This way of modelling the response does not produce very good error messages. See https://github.com/comit-network/comit-rs/pull/3262.